### PR TITLE
ASoC: SOF: IPC4: Add initial firmware handling support

### PR DESCRIPTION
**Pull request by ujfalusi**
Hi,

The series implements basic support for parsing the main firmware manifest for IPC4 based image.
Platform code must set up the `struct sof_ipc4_data` and sdev->private should be set to point to it since some parameters for parsing the firmware image is platform dependent.

### CODE REVIEW

File: sound/soc/sof/ipc4-priv.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause) */
```
Comment (plbossart) [inline sound/soc/sof/ipc4-priv.h:1]: commit message typo: "Add a struct'

File: sound/soc/sof/ipc4-priv.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+#include "sof-priv.h"
+
+/**
+ * struct sof_ipc4_private_data - IPC4-specific data
+ * @num_fw_modules : Number of modules in base FW
+ * @fw_modules: Array of base FW modules
+ * @manifest_fw_hdr_offset: FW header offset in the manifest
+ */
+struct sof_ipc4_data {
```
Comment (plbossart) [inline sound/soc/sof/ipc4-priv.h]: the name doesn't match the kernel-doc, and it's as vague as can be. 'data' could refer to just about anything.

sof_ipc4_fw_data?

File: sound/soc/sof/ipc4-priv.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+	int num_fw_modules;
+	void *fw_modules;
+	u32 manifest_fw_hdr_offset;
```
Comment (plbossart) [inline sound/soc/sof/ipc4-priv.h]: add this first?

**Review by plbossart — CHANGES_REQUESTED** [review]

File: sound/soc/sof/ipc4-priv.h (hunk 2 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-priv.h]: Oops, I have overlooked this.

Yes, `sof_ipc4_fw_data` is more appropriate

File: sound/soc/sof/ipc4-priv.h (hunk 3 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-priv.h]: Sure.

**Comment by ujfalusi** [comment]
Changes since v1:
- Typo fixed in commit messages (There were others than what @plbossart pointed out)
- Clarified some of the commit messages at the same time.
- sof_ipc4_data is renamed to sof_ipc4_fw_data and struct is re-arranged
- New patch is added to set up the sof_ipc4_fw_data for IPC4 capable platforms in case they are booted in IPC4 mode.

File: include/sound/sof/ext_manifest4.h [hunk 4]
```diff
@@ -0,0 +1,119 @@
+/* SPDX-License-Identifier: ((GPL-2.0 WITH Linux-syscall-note) OR BSD-3-Clause) */
```
Comment (ranj063) [inline include/sound/sof/ext_manifest4.h:1]: @ujfalusi missing other people's co-developed. Originally written by Bard and Rander

File: sound/soc/sof/ipc4-loader.c [hunk 5]
```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
```
Comment (ranj063) [inline sound/soc/sof/ipc4-loader.c:1]: this one too

**Comment by ujfalusi** [comment]
Changes since v2:
- Co-developed-by tags added

**Review by bardliao — APPROVED** [review]

**Review by RanderWang — APPROVED** [review]

**Review by plbossart — APPROVED** [review]

**Comment by plbossart** [comment]
let's leave a chance to @ranj063 to review/merge.

**Comment by ranj063** [comment]
@ujfalusi there are a couple of patches in this series without your sign-off. Since it is your PR, we wont be gettingyour reviewed-by tag either. So maybe we'll have to fix it when we checkerry-pick to add it later? I dont want to block merging for it

**Review by ranj063 — APPROVED** [review]

**Comment by ujfalusi** [comment]
> @ujfalusi there are a couple of patches in this series without your sign-off. Since it is your PR, we wont be gettingyour reviewed-by tag either. So maybe we'll have to fix it when we checkerry-pick to add it later? I dont want to block merging for it

Oops, well, if I will be sending it upstream they will get my SoB anyways. Sorry for missing that.

**Pull request merged**